sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{   
+
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter>&& exporter,
+                                       const int max_queue_size,
+                                       const int schedule_delay_millis,
+                                       const int max_export_batch_size)
+       :exporter_(std::move(exporter)),
+        max_queue_size_(max_queue_size), 
+        schedule_delay_millis_(schedule_delay_millis),
+        max_export_batch_size_(max_export_batch_size)   
+{  
+    buffer_ = std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+    
+    //Start the background worker thread
+    worker_thread_ = std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+
+} 
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept 
+{
+    return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept  
+{
+    // no-op
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept 
+{
+    if(is_shutdown_){
+        // TODO: log in glog
+        return;
+    }
+
+    if(static_cast<int>(buffer_->size()) == max_queue_size_){
+        // TODO: glog that spans will likely be dropped
+    }
```

Comment:
Would we still add spans past max_queue_size? Also, it is probably safer to compare using >= instead of == here.

Reply:
We would drop spans immediately.